File: leetcodejava/src/main/java/com/leetcode/Two.java
package com.leetcode;

import com.entity.ListNode;

/**
 * @author: John
 * @date: 2020-02-14 16:46
 * @description: 2
 **/
public class Two {
//    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
//        ListNode dummyHead = new ListNode(0);
//        ListNode p = l1, q = l2, curr = dummyHead;
//        int carry = 0;
//        while (p != null || q != null) {
//            int x = (p != null) ? p.val : 0;
//            int y = (q != null) ? q.val : 0;
//            int sum = carry + x + y;
//            carry = sum / 10;
//            curr.next = new ListNode(sum % 10);
//            curr = curr.next;
//            if (p != null) {
//                p = p.next;
//            }
//            if (q != null) {
//                q = q.next;
//            }
//        }
//        if  (carry > 0) {
//            curr.next = new ListNode(carry);
//        }
//        return dummyHead.next;
//    }


    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        int carry = 0;
        ListNode dummy1 = l1, dummy2 = l2;
        while ((l1 != null && l2 != null) || carry > 0) {
            int val1 = l1 == null ? 0 : l1.val;
            int val2 = l2 == null ? 0 : l2.val;
            int sum = val1 + val2 + carry;
            carry = sum / 10;
//            l1.val = l2.val = sum % 10;
            if (l1 != null) {
                l1.val = sum % 10;
            }
            if (l2 != null) {
                l2.val = sum % 10;
            }
            l1 = l1 == null ? null : l1.next;
            l2 = l2 == null ? null : l2.next;
        }
        if (l1 == null) {
            return dummy2;
        }
        return dummy1;
    }

    public static void main(String[] args) {
        ListNode l1 = new ListNode(9);
        ListNode l11 = new ListNode(9);
        ListNode l111 = new ListNode(9);
        ListNode l1111 = new ListNode(9);
        ListNode l11111 = new ListNode(9);
        ListNode l111111 = new ListNode(9);
        ListNode l1111111 = new ListNode(9);
        l111111.next = l1111111;
        l11111.next = l111111;
        l1111.next = l11111;
        l111.next = l1111;
        l11.next = l111;
        l1.next = l11;

        ListNode l2 = new ListNode(9);
        ListNode l22 = new ListNode(9);
        ListNode l222 = new ListNode(9);
        ListNode l2222 = new ListNode(9);
//        ListNode l222 = new ListNode(4);
        l222.next = l2222;
        l22.next = l222;
        l2.next = l22;

        Two two = new Two();
        ListNode a = two.addTwoNumbers(l1, l2);
    }
}

File: leetcodecpp/leetcode/Two.cpp
//
// Created by John on 2022/4/9.
//
#include "../algorithm/LeetCodeInterview.h"

using namespace std;

class Solution {
public:
    static
//    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
//        int carry = 0;
//        ListNode dummyNode = ListNode();
//        ListNode *curr = &dummyNode;
//
//        while (l1 != nullptr || l2 != nullptr) {
//            int total = (l1 == nullptr ? 0 : l1->val) + (l2 == nullptr ? 0 : l2->val) + carry;
//            curr->next = new ListNode(total % 10);
//            curr = curr->next;
//            carry = total / 10;
//            if (l1 != nullptr) {
//                l1 = l1->next;
//            }
//            if (l2 != nullptr) {
//                l2 = l2->next;
//            }
//        }
//        if (carry > 0) {
//            curr->next = new ListNode(carry);
//        }
//        return dummyNode.next;
//    }

//    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
//        int c = 0;
//        ListNode newHead(0);
//        ListNode *t = &newHead;
//        while (c || l1 || l2) {
//            c += (l1 ? l1->val : 0) + (l2 ? l2->val : 0);
//            t->next = new ListNode(c % 10);
//            t = t->next;
//            c /= 10;
//            if (l1) {
//                l1 = l1->next;
//            }
//            if (l2) {
//                l2 = l2->next;
//            }
//        }
//        return t->next;
//    }

//    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
//        if(!l1 && !l2) return NULL;
//        int c = (l1? l1->val:0) + (l2? l2->val:0);
//        ListNode *newHead = new ListNode(c%10), *next = l1? l1->next:NULL;
//        c /= 10;
//        if(next) next->val += c;
//        else if(c) next = new ListNode(c);
//        newHead->next = addTwoNumbers(l2? l2->next:NULL, next);
//        return newHead;
//    }
    ListNode *addTwoNumbers(ListNode *l1, ListNode *l2) {
        // if both all nullptr return NULL
        if (!l1 && !l2) {
            return NULL;
        }
        int c = (l1 ? l1->val : 0) + (l2 ? l2->val : 0);
        ListNode *newHead = new ListNode(c%10), *next = l1 ? l1->next : NULL;
        c /= 10;
        if (next) {
            next->val += c;
        } else if (c) {
            next = new ListNode(c);
        }
        newHead->next = addTwoNumbers(l2 ? l2->next : NULL, next);
        return newHead;
    }


//public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
//        ListNode dummyHead = new ListNode(0);
//        ListNode p = l1, q = l2, curr = dummyHead;
//        int carry = 0;
//        while (p != null || q != null) {
//            int x = (p != null) ? p.val : 0;
//            int y = (q != null) ? q.val : 0;
//            int sum = carry + x + y;
//            carry = sum / 10;
//            curr.next = new ListNode(sum % 10);
//            curr = curr.next;
//            if (p != null) p = p.next;
//            if (q != null) q = q.next;
//        }
//        if (carry > 0) {
//            curr.next = new ListNode(carry);
//        }
//        return dummyHead.next;
//    }

};


